File: java/AbstractFactory/src/absfactory/USPhoneNumber.java
package absfactory;

public class USPhoneNumber implements PhoneNumber{
	String phoneNumber;
	
	public USPhoneNumber(String phoneNumber) {
		if(!isValidPhoneNumber(phoneNumber)) {
			throw new IllegalArgumentException("Invalid US phone number");
		}
		this.phoneNumber = phoneNumber;
	}
	@Override
	public boolean isValidPhoneNumber(String phoneNumber) {
		return phoneNumber.matches("^(\\([0-9]{3}\\) |[0-9]{3}-)[0-9]{3}-[0-9]{4}$");
	}

}

File: ruby/AbstractFactory/us_phone_number.rb
require_relative 'phonenumber'

class USPhoneNumber < PhoneNumber
  def initialize(phone_number)
    if is_valid_phonenumber(phone_number)
      phone_number = phone_number
    else
      raise 'Invalid post code'
    end

    puts "Creating a instance of USPhoneNumber #{phone_number}"
  end

  def is_valid_phonenumber(phone_number)
    (phone_number =~ /^(\\([0-9]{3}\\) |[0-9]{3}-)[0-9]{3}-[0-9]{4}$/)
  end
end